### ISSUE TITLE
Use torch.nn.DataParallel for intra-node computation

### ISSUE THREAD

**Issue by IamTao** [mlbench/mlbench-core#46]
It might be a good choice to use `torch.nn.DataParallel` for intra-node computation (across multi-GPUs) and intra-node gradient aggregation, and then use different communication backends for inter-node communication.

**Comment by ehoelzl** [comment]
Wouldn't it be more appropriate to use `DistributedDataParallel`, as referenced [here](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#comparison-between-dataparallel-and-distributeddataparallel)

**Comment by IamTao** [comment]
I think it should be a standard baseline for us to compare with, and we should check both of `DataParallel ` and `DistributedDataParallel` for the intra-node case. But indeed I am not sure if we can directly use `DistributedDataParallel` in our current framework for intra-node communication (it also requires to run `init_process_group`).

BTW, I think the `DistributedDataParallel` is mainly designed/optimized for centralized training across multi-nodes but it loses the flexibility to use different communication strategies (e.g. compressed gradient), different communication topologies (e.g. ring topology for decentralized training). It is good to get the results from `DistributedDataParallel` (for distributed training) and compares it with our sync scheme.

**Comment by martinjaggi** [comment]
DDP overlaps gradient computation with the communication. the effect should be noticeable. how does it compare to our reference implementations?

**Comment by ehoelzl** [comment]
Closing in favor of https://github.com/mlbench/mlbench-benchmarks/issues/69
